src/stdlib/table.rs

```diff
@@ -731,3 +759,130 @@ fn array_insert_shift<'gc>(
     let length = table.length() as usize;
     (inner(table, length, key, value), length)
 }
+
+#[derive(Debug, Copy, Clone, Error)]
+pub enum TableConcatError {
+    #[error("resulting string too long in table.concat")]
+    Overflow,
+}
+
+fn concat_impl<'gc>(ctx: Context<'gc>, sep: Option<Value<'gc>>) -> BoxSequence<'gc> {
```

Comment:
This has a lot of repeated logic with `meta_ops::concat`. I might be missing something, but are they identical except for the inclusion of an optional separator (and start / end index?)

This may sound heretical, but what if we implemented this by *interspersing* the separator into the stack and then calling `meta_ops::concat` so as not to duplicate everything?

Reply:
The start/end index is handled by `table.unpack`, so besides for the separator, the logic is the same for both.  I'm not a fan of interspersing the separator because it doubles the intermediate memory usage, which could be a problem in reasonable use-cases.  (For example, joining an large array of integers by commas -- the stack manipulation would likely be the most expensive part.)

Another approach would be to just have an implementation that supports separators and have both the meta-op and `table.concat` use that internally, but that adds some complexity to the normal concat case.

Reply:
Oh, another minor difference is that for the concat metaop, the first half isn't a Sequence, but it has to be in the `table.concat` case, to be able to tail-call `table.unpack` and have it return into the concat implementation.  As such, some of the inner implementation can be reused (and some currently is), but each case has different control-flow requirements which make merging them difficult.